ext/bcmath/bcmath.c

```diff
@@ -174,15 +177,23 @@ PHP_FUNCTION(bcadd)
 	bc_init_num(&first);
 	bc_init_num(&second);
 	bc_init_num(&result);
-	php_str2num(&first, ZSTR_VAL(left));
-	php_str2num(&second, ZSTR_VAL(right));
+
+	if (php_str2num(&first, ZSTR_VAL(left)) == FAILURE) {
+		zend_argument_value_error(1, "bcmath function argument is not well-formed");
+		RETURN_THROWS();
```

Comment:
Replace `RETURN_THROWS()` with `goto cleanup;` and then add `cleanup:` before the three `bc_free_num` call at the end of the function. Similar for the other cases.

You should also be able to add a test that currently leaks, e.g. using `bcadd("1", "a");` i.e. where the first parameter is correct but the second one isn't.

Reply:
@nikic Thanks, I added the `goto`'s and added a test. I'm wondering why a memory leak only occurs when the second parameter is invalid and not when the first parameter is invalid?

Reply:
Presumably bc_init_num by itself doesn't allocate anything, so we need to the first number to be initialized to something non-trivial in order to leak.

Reply:
`bc_init_num` copies (i.e. increases the ref count of) a module global. I think there is a theoretical leak on GSHUTDOWN.